### PR TITLE
enable FSDP2 + fp8 all-gather and fix TP fp8 all-gather

### PR DESCRIPTION
**Pull request by weifengpy**
we have landed fp8 all-gather optimizations in float8_experimental https://github.com/pytorch-labs/float8_experimental/pull/266/

this PR proposes torchtitan changes. also include fp8 in CI
```
from float8_experimental.fsdp_utils import precompute_float8_dynamic_scale_for_fsdp
# inside the training loop
model(input).sum().backward()
optim.step()
precompute_float8_dynamic_scale_for_fsdp(model)
```

FSDP2 fp8 all-gather are added to CI
```
CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --training.enable_fp8_linear
CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --training.enable_fp8_linear --training.enable_fsdp_fp8_all_gather
CONFIG_FILE="./train_configs/debug_model.toml" ./run_llama_train.sh --training.enable_fp8_linear --training.enable_fsdp_fp8_all_gather --training.precompute_float8_dynamic_scale_for_fsdp
```

TP fp8 all-gather are locally tested. will add them to CI after uploading a new tokenizer with vacab size 2560 (divisible by 16)
```
CONFIG_FILE="./train_configs/llama3_8b.toml" NGPU=4 ./run_llama_train.sh --training.enable_fp8_linear --training.data_parallel_degree 1 --training.tensor_parallel_degree 4
CONFIG_FILE="./train_configs/llama3_8b.toml" NGPU=4 ./run_llama_train.sh --training.enable_fp8_linear --training.data_parallel_degree 2 --training.tensor_parallel_degree 2
```

precompute scales after optimizer.step
Screenshot 2024-07-12 at 5 11 14 PM

FSDP2 pre-all-gather do not have any small all-reduces
Screenshot 2024-07-12 at 5 13 04 PM

TODO
* upload tokenizer with vacab size 2560 to enable CI on TP fp8 all-gather
* torch.compile complains about fp8
* add delayed scaling and brainstorm about best config option to express fp8
* compare perf between delayed scaling and dynamic scaling https://github.com/pytorch-labs/float8_experimental/pull/312/files

### CODE REVIEW

File: test_runner.py [hunk 1]
```diff
@@ -273,6 +273,39 @@ def build_test_list():
             "fsdp2_mem_tracker",
             ngpu=4,
         ),
+        OverrideDefinitions(
```
Comment (weifengpy) [inline test_runner.py:276]: added followings to CI
* 1D fsdp original dtype all-gather
* 1D fsdp fp8 all-gather
* 1D fsdp fp8 all-gather with precomputed dynamic scales

need follow ups to enable TP fp8 all-gather in CI: current CI tokenizer has 2556, not divisible by 16) https://github.com/pytorch/torchtitan/issues/461
* 1D TP fp8 all-gather
* 2D FSDP + TP fp8 all-gather

File: test_runner.py [hunk 2]
```diff
@@ -273,6 +273,39 @@ def build_test_list():
...
+            [
+                [
+                    "--training.fp8_linear",
+                ]
+            ],
+            "FSDP2 with bf16 all-gather",
```
Comment (tianyu-l) [inline test_runner.py]: maybe noob question: does all-gather always happen in bf16, or is it determined by `param_dtype` of FSDP2

File: test_runner.py [hunk 3]
```diff
@@ -273,6 +273,39 @@ def build_test_list():
...
+            "fp8_fsdp2_bf16_all_gather",
+            ngpu=4,
+        ),
+        OverrideDefinitions(
+            [
+                [
+                    "--training.fp8_linear",
+                    "--training.enable_fsdp_fp8_all_gather",
```
Comment (tianyu-l) [inline test_runner.py:290]: should it only be enabled when `fp8_linear` and FSDP2 are used?
shall we add an assertion somewhere?

File: test_runner.py [hunk 4]
```diff
@@ -273,6 +273,39 @@ def build_test_list():
...
+            "FSDP2 with fp8 all-gather",
+            "fp8_fsdp2_fp8_all_gather",
+            ngpu=4,
+        ),
+        OverrideDefinitions(
+            [
+                [
+                    "--training.fp8_linear",
+                    "--training.enable_fsdp_fp8_all_gather",
+                    "--training.precompute_float8_dynamic_scale_for_fsdp",
```
Comment (tianyu-l) [inline test_runner.py:302]: same question as above.

File: train.py [hunk 5]
```diff
@@ -398,6 +399,9 @@ def loss_fn(pred, labels):
             optimizers.step()
             lr_schedulers.step()
 
+            if job_config.training.precompute_float8_dynamic_scale_for_fsdp:
+                precompute_float8_dynamic_scale_for_fsdp(model)
```
Comment (tianyu-l) [inline train.py:409]: maybe a noob question: could you briefly explain what this is doing?
I wonder since we are already using context functions for FP8, can we have a context and run it in a `.step()` function here, just like optimizer, lr scheduler, and profiler. This would make the code consistent.

File: test_runner.py (hunk 2 above)
Comment (weifengpy) [reply test_runner.py]: good question! I should change this to "all-gather in original dtype"

when mixed_precision is turned off (no `param_dtype` configed), FSDP2 communicate in model's original dtype

when mixed_precision is turned on (`param_dtype`=`xxx`), FSDP2 communicates according to `param_dtype`

File: test_runner.py (hunk 3 above)
Comment (weifengpy) [reply test_runner.py:290]: good point. so the logic is like follows. if `fp8_linear=False`, `enable_fsdp_fp8_all_gather` won't be executed regardless of true or false. do you prefer to move it to top-level?  

```
if fp8_linear:
   if enable_fsdp_fp8_all_gather
```

File: test_runner.py (hunk 4 above)
Comment (weifengpy) [reply test_runner.py:302]: same to situation to `enable_fsdp_fp8_all_gather`

File: train.py (hunk 5 above)
Comment (weifengpy) [reply train.py:409]: > could you briefly explain what this is doing

`precompute_float8_dynamic_scale_for_fsdp` is a for-loop over `model.parameters()`. it issues a single all-reduce for all parameters, ie `abs(max(param)) for param in model.parameters()` and save amax/scale as `param._precomputed_scale`. this speed up the training loop since we do not need to compute amax/scale for each parameters in the training loop

> we are already using context functions for FP8

do you refer to `set_enable_fsdp_fp8_all_gather` ? That's for model intiaitialization where we swap `nn.Linear` with user-defined float8 linear. `precompute_float8_dynamic_scale_for_fsdp` is for training loop

Comment (weifengpy) [reply train.py:409]: per suggestion, raise error if `use_fp8_linear=False` or `enable_fsdp_fp8_all_gather =False`

File: test_runner.py (hunk 4 above)
Comment (weifengpy) [reply test_runner.py:302]: I changed the code to raise ValueError if `fp8_linear=False` or `enable_fsdp_fp8_all_gather=False` when `precompute_float8_dynamic_scale_for_fsdp=True`

File: train.py (hunk 5 above)
Comment (yifuwang) [reply train.py:409]: noob q: do we eventually want to just put this in fsdp2?

Comment (awgu) [reply train.py:409]: It has to be done after optimizer step (since parameter values change). Are you suggesting to run this in the root module's pre-forward?

Comment (yifuwang) [reply train.py:409]: Yeah anywhere between the n-1th optimizer step and the first all-gather in the nth step where fsdp2 has control (if there's any).

Comment (awgu) [reply train.py:409]: That makes sense. I think one concern is that FSDP is agnostic to the fp8 all-gather. FSDP does not know that the `fsdp_pre_all_gather` and `fsdp_post_all_gather` of the `Float8Linear.weight`s are implemented to do fp8 all-gather, so at best, the user still would need to register a module forward pre-hook or something to run this method.

Comment (yifuwang) [reply train.py:409]: Ah I see. Somehow I thought fsdp2 was fp8-aware

File: train.py [hunk 6]
```diff
@@ -398,6 +404,17 @@ def loss_fn(pred, labels):
             optimizers.step()
             lr_schedulers.step()
```
Comment (weifengpy) [inline train.py:401]: add comment to explain `precompute_float8_dynamic_scale_for_fsdp`

File: torchtitan/config_manager.py [hunk 7]
```diff
@@ -347,6 +347,18 @@ def __init__(self):
                 here: https://github.com/pytorch-labs/float8_experimental
             """,
         )
+        self.parser.add_argument(
```
Comment (tianyu-l) [inline torchtitan/config_manager.py:350]: As discussed offline, let's refactor fp8 configs, e.g. have a dedicated field for enabling fp8 or not.

File: torchtitan/float8_linear.py [hunk 8]
```diff
@@ -27,8 +44,8 @@ def build_fp8_linear(model: nn.Module, job_config: JobConfig):
     This will mutate the model inplace.
     """
     use_fp8_linear = job_config.training.fp8_linear
+    enable_fsdp_fp8_all_gather = job_config.training.enable_fsdp_fp8_all_gather
```
Comment (tianyu-l) [inline torchtitan/float8_linear.py]: discussed offline: please check if it makes sense to enable it only when `dp_degree > 1`

File: train.py [hunk 9]
```diff
@@ -218,6 +219,11 @@ def loss_fn(pred, labels):
     # apply fp8 linear module swap
     if job_config.training.fp8_linear:
         build_fp8_linear(whole_model, job_config)
+    else:
```
Comment (tianyu-l) [inline train.py]: can remove this to favor simplicity if it is a no-op flag when `fp8_linear=False`

File: train.py [hunk 10]
```diff
@@ -398,6 +404,17 @@ def loss_fn(pred, labels):
             optimizers.step()
             lr_schedulers.step()
 
+            if job_config.training.precompute_float8_dynamic_scale_for_fsdp:
```
Comment (tianyu-l) [inline train.py]: discussed offline: can refactor to make it simpler

**Review by tianyu-l — COMMENTED** [review]
Left some comments to record what we discussed offline.

File: torchtitan/config_manager.py (hunk 7 above)
Comment (weifengpy) [reply torchtitan/config_manager.py:350]: renamed `fp8_linear` to `enable_fp8_linear`

File: torchtitan/parallelisms/parallelize_llama.py [hunk 11]
```diff
@@ -117,12 +117,22 @@ def selective_checkpointing_context_fn():
 
 def get_tp_parallel_strategy(
     job_config: JobConfig,
+    model: nn.Module,
 ) -> Tuple[RowwiseParallel, ColwiseParallel, PrepareModuleInput]:
     """Get the parallel strategy for the transformer model.
 
     This function handles the special case of using float8 with tensor parallelism.
     """
-    if job_config.training.fp8_linear == "dynamic":
```
Comment (weifengpy) [inline torchtitan/parallelisms/parallelize_llama.py:125]: `fp8_linear == "dynamic"` is outdated after recent unification of dynamic/delayed scaling (`Float8Linear`)
https://github.com/pytorch/torchtitan/pull/436 

update it in this PR to make TP fp8 all-gather work again

EDIT: Will enable TP in CI to prevention after having a new tokenizer with vacab size 2560

File: train.py (hunk 6 above)
Comment (weifengpy) [reply train.py:401]: done

File: train.py (hunk 10 above)
Comment (weifengpy) [reply train.py]: removed `ValueError` when `enable_fp8_linear/enable_fsdp_fp8_all_gather=False`

File: train.py (hunk 9 above)
Comment (weifengpy) [reply train.py]: removed `ValueError` on `enable_fp8_linear=False`

File: torchtitan/float8_linear.py (hunk 8 above)
Comment (weifengpy) [reply torchtitan/float8_linear.py]: added check on `parallel_dims.dp_enabled`

File: torchtitan/config_manager.py (hunk 7 above)
Comment (drisspg) [reply torchtitan/config_manager.py:350]: I think one thing to note is that right now this is a boolean which will swap to the default float8 recipe
Dynamic scaling x Tensor wise ScalingGranularity x all tensors involved in the matmul [ input, weight, grad]

I think we should brainstorm on an elegant solutions for users to express their desired config here

File: torchtitan/float8_linear.py [hunk 12]
```diff
@@ -12,31 +12,59 @@
...
+        job_config.training.enable_fsdp_fp8_all_gather and dp_enabled
+    )
     try:
-        from float8_experimental.float8_linear import Float8Linear
         from float8_experimental.float8_linear_utils import (
             swap_linear_with_float8_linear,
         )
+
+        # Mutates the model inplace replacing instances of torch.nn.Linear with Float8Linear
+        with set_enable_fsdp_fp8_all_gather(enable_fsdp_fp8_all_gather):
```
Comment (drisspg) [inline torchtitan/float8_linear.py:59]: noop Q: do we need this in a context manager to make testing + resetting easier?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 13]
```diff
@@ -117,12 +117,22 @@ def selective_checkpointing_context_fn():
...
+    if job_config.training.enable_fp8_linear:
+        from float8_experimental.float8_linear import Float8Linear, TensorScalingType
+
+        if any(
+            isinstance(m, Float8Linear)
+            and m.scaling_type_w is TensorScalingType.DELAYED
+            for m in module.modules()
+        ):
+            raise NotImplementedError("Only supports delayed scaling")
```
Comment (drisspg) [inline torchtitan/parallelisms/parallelize_llama.py]: Should be "only supports dynamic scaling" right?

Comment (weifengpy) [reply torchtitan/parallelisms/parallelize_llama.py]: good catch! will change it

File: torchtitan/float8_linear.py (hunk 12 above)
Comment (weifengpy) [reply torchtitan/float8_linear.py:59]: hmm. `set_enable_fsdp_fp8_all_gather` is a context manager right now. do you mean "why" it should be a context manager ? 

EDIT: I also see you mentioned "make testing + resetting easier", which answered why. so I am not sure  if it's a question for me

File: test_runner.py [hunk 14]
```diff
@@ -273,6 +273,64 @@ def build_test_list():
...
+        OverrideDefinitions(
+            [
+                [
+                    "--training.enable_fp8_linear",
+                    "--training.enable_fsdp_fp8_all_gather",
+                    "--training.precompute_float8_dynamic_scale_for_fsdp",
+                    "--training.tensor_parallel_degree 2",
+                ]
+            ],
+            "FSDP2 with fp8 all-gather and precomputed dynamic scales",
```
Comment (tianyu-l) [inline test_runner.py]: nit: comment for 2D

Comment (weifengpy) [reply test_runner.py]: in the end I have to remove 2D from this PR. current CI tokenizer has vacab size = 2556. However, fp8 gemm need the vacab size to be divisible by 16 https://github.com/pytorch/torchtitan/issues/461

I can follow up with you on how to have a tokenizer with vacab size = 2560 to unblock 1D TP + fp8, and 2D + fp8 in CI

File: torchtitan/config_manager.py (hunk 7 above)
Comment (weifengpy) [reply torchtitan/config_manager.py:350]: good question. evetually we might have to expose args/kwargs from `swap_linear_with_float8_linear` for flexibility

File: torchtitan/float8_linear.py [hunk 15]
```diff
@@ -12,31 +12,60 @@
...
+):
     """
     This function converts the linear layers to `Float8Linear`. Note that today,
     only dynamic tensor scaling (the default) is supported.
 
     This will mutate the model inplace.
     """
-    use_fp8_linear = job_config.training.fp8_linear
+    enable_fp8_linear = job_config.training.enable_fp8_linear
+    if not enable_fp8_linear:
```
Comment (tianyu-l) [inline torchtitan/float8_linear.py]: let's remove this redundant check

**Review by tianyu-l — APPROVED** [review]
The PR looks good to me, except that there is an ongoing discussion on what's the best way to support fp8 GEMM on tensors of irregular shapes (when not divisible by 16) e.g. coming from TP + test tokenizer model in this repo (see #461). Maybe we should allow irregular shapes but with potential perf regression. Will need to discuss more on this later.

File: train.py [hunk 16]
```diff
@@ -19,6 +19,7 @@
 
 import torch
 import torch.nn.functional as F
+from float8_experimental.fsdp_utils import precompute_float8_dynamic_scale_for_fsdp
```
Comment (wanchaol) [inline train.py:22]: @weifengpy I think we should hide this import to the path where `enable_fp8_allgather` path happened? 

The problem here is that for every feature that requires an additional install from other dependency, we should try to hide the import to the path that uses it instead of import it globally, otherwise for users who didn't install the float8_experimental, if they rebase, and it would just fail to train for them.

Please submit a follow up PR to fix this

Comment (weifengpy) [reply train.py:22]: got you. I am moving it from top-level to if-else now  https://github.com/pytorch/torchtitan/pull/464

thanks for the timely reminder
